### PR TITLE
ZkClient related improvments

### PR DESCRIPTION
**Pull request by jiajunwang**
We identify 2 potential issues that may cause a retrying ZK operation failed unexpectedly. These commits fix the problem.

### CODE REVIEW

**Comment by dasahcc** [comment]
LGTM! Make sure you pass all the tests.

File: helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java [hunk 1]
```diff
@@ -674,12 +684,24 @@ public void handleChildChange(String parentPath, List<String> currentChilds) {
   /**
    * Invoke the listener for the last time so that the listener could clean up resources
    */
+  @Deprecated
   public void reset() {
+    reset(true);
+  }
+
+  void reset(boolean isShutdown) {
     logger.info("Resetting CallbackHandler: " + this.toString());
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java]: include isShutdown here in the log?

**Comment by jiajunwang** [comment]
I split the ZkConnection commit to separate logic change and file movement. Just to ensure the history of change is clear.

File: helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkConnection.java [hunk 2]
```diff
@@ -84,6 +83,25 @@ public void close() throws InterruptedException {
...
 
+  protected void reconnect(Watcher watcher) throws InterruptedException {
+    _zookeeperLock.lock();
+    try {
+      if (_zk == null) {
+        throw new IllegalStateException("zk client has not been connected or already been closed");
+      }
+      ZooKeeper prevZk = _zk;
+      try {
+        LOG.debug("Creating new ZookKeeper instance to reconnect to " + _servers + ".");
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkConnection.java]: LOG.info?

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkConnection.java]: We already have state change logs. In addition, if the reconnect fails, we will print warning log in the ZkClient. Here, to be consistent with other ZkConnection, I think it's better that we keep it a debug log.

**Pull request merged**